torchdrivesim/traffic_lights.py

```diff
@@ -0,0 +1,120 @@
+import json
+import random
+from typing import List
+from functools import reduce, lru_cache
+from enum import auto, Enum
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+
+class ControlStateType(Enum):
+    none = auto()
+    green = auto()
+    yellow = auto()
+    red = auto()
+
+
+class TrafficLightStateMachine:
+    def __init__(self, json_file_path: str, tick_to_second_ratio: float = 0.1):
+        self._tick_to_second_ratio = tick_to_second_ratio
+        self._states = self.load_from_json(json_file_path)
+        for state in self._states:
+            state['duration'] = float(state['duration'])
+        self._time_remaining, self._current_state, self._duration = None, None, None
+        self.reset()
+
+    def load_from_json(self, json_file_path: str):
+        with open(json_file_path, 'rb') as f:
+            light_state_machine = json.load(f)
+        return light_state_machine
+
+    def reset(self):
+        state = random.randint(0, len(self._states) - 1)
+        self.set_to(state, self._states[state]['duration'])
+
+    def set_to(self, state_index: int, time_remaining: float):
+        state = state_index
+        if state_index < 0:
+            state = 0
+        elif state_index >= len(self._states):
+            state = len(self._states) - 1
+        self._current_state = self._states[state]
+        self._duration = self._current_state['duration']
+        self._time_remaining = time_remaining if time_remaining <= self._duration else self._duration
+
+    def tick(self, dt: float):
+        self._time_remaining -= (dt * self._tick_to_second_ratio)
```

Comment:
This `_tick_to_second_ratio` is very confusing to me. I would expect `dt` to already be in seconds, and the state duration to be specified in seconds as well, so that we can be timestep agnostic.